diff --git a/sweph.c b/sweph.c
index caa82a5..f1ba7f1 100644
--- a/sweph.c
+++ b/sweph.c
@@ -248,8 +248,10 @@ HANDLE dllhandle = NULL;        // global used in swe_version
 #ifndef __USE_GNU
 #define __USE_GNU
 #endif
-#include <dlfcn.h>		// must be linked with -ldl
-  static Dl_info dli;
+#if !defined(__WASI__)
+	#include <dlfcn.h>		// must be linked with -ldl
+		static Dl_info dli;
+#endif
 #endif
 #endif // MSDOS
 
@@ -258,29 +260,31 @@ char *CALL_CONV swe_get_library_path(char *s)
   size_t bytes;
   size_t len;
   *s = '\0';
-#if !defined(__APPLE) 
-  len = AS_MAXCH;
-  bytes = 0;
-#if MSDOS
-  bytes = GetModuleFileName((HMODULE) dllhandle, (TCHAR*) s, (DWORD) len);
-#else
-  #ifdef __GNUC__
-    if (dladdr((void *)swe_version, &dli) != 0) {
-      if (strlen(dli.dli_fname) >= len) {
-	strncpy(s, dli.dli_fname, len);
-	s[len] = '\0';
-      } else{
-	strcpy(s, dli.dli_fname);
+#if !defined(__WASI__)
+  #if !defined(__APPLE) 
+    len = AS_MAXCH;
+    bytes = 0;
+  #if MSDOS
+    bytes = GetModuleFileName((HMODULE) dllhandle, (TCHAR*) s, (DWORD) len);
+  #else
+    #ifdef __GNUC__
+      if (dladdr((void *)swe_version, &dli) != 0) {
+        if (strlen(dli.dli_fname) >= len) {
+    strncpy(s, dli.dli_fname, len);
+    s[len] = '\0';
+        } else{
+    strcpy(s, dli.dli_fname);
+        }
+        bytes = strlen(s);
+      } else {
+        bytes = readlink("/proc/self/exe", s, len);
       }
-      bytes = strlen(s);
-    } else {
+    #else
       bytes = readlink("/proc/self/exe", s, len);
-    }
-  #else
-    bytes = readlink("/proc/self/exe", s, len);
+    #endif
+  #endif
+    s[bytes] = '\0';
   #endif
-#endif
-  s[bytes] = '\0';
 #endif
   return s;
 }
diff --git a/swetest.c b/swetest.c
index aa92659..3f83f5c 100644
--- a/swetest.c
+++ b/swetest.c
@@ -1277,10 +1277,12 @@ int main(int argc, char *argv[])
   }
   // if (! with_header && ! has_n)
   //  with_header = TRUE;
-#if HPUNIX
-  gethostname (hostname, 80);
-  if (strstr(hostname, "as10") != NULL) 
-    line_limit = 1000;
+#if !defined(__WASI__)
+  #if HPUNIX
+    gethostname (hostname, 80);
+    if (strstr(hostname, "as10") != NULL) 
+      line_limit = 1000;
+  #endif
 #endif
 #if MSDOS
   SetConsoleOutputCP(65001);	// set console to utf-8,
